### DISCUSSION TITLE
How to switch Git LFS server on the same remote

### DISCUSSION THREAD

**Discussion by lucasrangit** [git-lfs/git-lfs#4984] [Q&A]
I need to migrate from GitHub LFS to JFrog LFS client side using `git lfs fetch` and `git lfs push` but I am confused. Even after the recent improvements to `git-lfs-push(1)` in d16c3d6 by @chrisd8088 and the comment by @bk2204  https://github.com/git-lfs/git-lfs/issues/3984#issuecomment-574410503 . 

> Barring that, you'd want to do a full `git clone` as step 0, do `git lfs fetch --all` between steps 1 and 2, and then do a `git lfs push --all` after step 6. That should ensure that all of the objects are copied across between the two servers.

1. Does a "full `git clone`" mean a `git clone --mirror` in order to get all remote branch (`refs/heads/*`) and tag (`refs/tags/*`) refs?

```
--all:
    This pushes all objects to the remote that are referenced by any commit
    reachable from the refs provided as arguments. If no refs are provided, then
    all local refs are pushed.  Note that this behavior differs from that of
    git-lfs-fetch(1) when its `--all` option is used; in that case, all refs
    are fetched, including refs other than those under `refs/heads` and
    `refs/tags`.  If you are migrating a repository with these commands, make
    sure to run `git lfs push` for any additional remote refs that contain
    Git LFS objects not reachable from your local refs.
```

2. With or without `--all`, it appears that only *local* refs are considered. In other words, when `git lfs push --all` is used, then all refs listed by `git show-ref` are checked for LFS objects that might need to be pushed. And when `git lfs push` is used, then only the current checkout branch ref is used (e.g. `refs/heads/my-git-checkout-branch`, ). Can you give an example how to _run `git lfs push` for any additional remote refs that contain Git LFS objects not reachable from your local refs_? For example, I would run `git lfs push origin remote-branch-that-do-not-locally-track-with-lfs-objects` if I never fetched that branch?

 I think the confusion is `git lfs fetch --all` considered both remote and local refs but `git lfs push --all` only considers local refs. And would `git fetch --all` be enough to get all remote refs or do I need to use `git clone --mirror`?

Thank you the for the help.

**Comment by chrisd8088** [comment]
Above all, I would strongly suggest experimenting and testing to ensure that whatever series of commands you use when migrating, you get the results you want.  That's always important since your situation may not fit the usual advice.

1. I suspect what @bk2204 meant by "full `git clone`" was not to use anything like a partial clone or a sparse checkout.  That said, using `git clone --bare` might be what you want.  (Note that `--mirror` implies `--bare`, and `--bare` will create `refs/heads` entries for branch heads at the remote, without mapping them to `refs/remotes/origin`.  Adding in the extra functionality of `--mirror` is probably not necessary.)

2. Running `git lfs push origin remote-branch-that-do-not-locally-track-with-lfs-objects` should work, I believe.  One thing you can do afterwards is check whether any Git LFS objects are missing from your migrated repo.  Make a clone of that repo (not your original source one, but the one you've pushed with `git lfs push`) and then in that run something like:
    ```
    git for-each-ref --shell --format="ref=%(refname)" | \
      while read entry; do eval "$entry"; printf "$ref\n"; git lfs fsck "$ref"; printf "\n"; done
    ```

**Reply by lucasrangit** [reply]
Thanks for the tip to check for missing LFS objects.

So to recap, the correct process to switch LFS repositories should be the following.

1. Run `git clone --bare` . Check that no "_remote refs that contain Git LFS objects not reachable from your local refs_" are listed, in other words no `refs/origin/remote/*` and instead all branches are under `refs/heads/*`. 
2. Run `git config --add lfs.url <url>` to set the new LFS url.
3. Run `git lfs push --all origin` to upload all objects for all local refs (which is all remote refs thanks using `git clone --bare`).

I am running that now. What I see is the progress update going to 100% very quickly and then dropping again as more files are added to the total. So other than the total size and transfer rate it's hard to tell that it's working.

**Reply by shoelzer** [reply]
Thanks, I was looking for a simple recipe like this. But after cloning don't you need to run `git lfs fetch --all` to make sure you have all the LFS files?
